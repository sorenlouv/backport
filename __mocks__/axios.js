const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
