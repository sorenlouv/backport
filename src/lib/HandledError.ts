export class HandledError extends Error {
  constructor(message: string) {
    super(message);
    Error.captureStackTrace(this, HandledError);
    this.name = 'HandledError';
  }
}

export function printHandledError(e: HandledError, { rethrow = true } = {}) {
  if (e.name === 'HandledError') {
    console.error(e.message);
  } else {
    console.error(e);
    if (rethrow) {
      throw e;
    }
  }
}
