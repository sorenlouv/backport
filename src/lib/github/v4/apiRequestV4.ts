import { DocumentNode } from 'graphql';
import { GraphQLClient, ClientError } from 'graphql-request';
import { getSdk } from '../../../graphql/generated';

export function getV4Client({
  githubApiBaseUrlV4,
  accessToken,
}: {
  githubApiBaseUrlV4?: string;
  accessToken: string;
}) {
  const client = new GraphQLClient(
    githubApiBaseUrlV4 ?? 'https://api.github.com/graphql',
    { headers: { Authorization: `bearer ${accessToken}` } },
  );
  return getSdk(client, async (action, operationName) => {
    try {
      return await action();
    } catch (err: unknown) {
      if (err instanceof ClientError) {
        throw new GithubV4Exception(err, `Error in ${operationName}`);
      }
      throw err;
    }
  });
}

export class GithubV4Exception extends Error {
  public readonly response: ClientError['response'];

  constructor(err: ClientError, contextMessage?: string) {
    const msgs =
      err.response.errors?.map((e) => e.message).join(', ') ||
      err.message ||
      'Unknown GitHub API v4 error';

    super(`${contextMessage ?? msgs} (GitHub API v4)`);
    this.name = 'GithubV4Exception';
    this.response = err.response;
  }
}

export function getQueryName(query: DocumentNode): string | undefined {
  if (query.definitions[0].kind === 'OperationDefinition') {
    return query.definitions[0].name?.value;
  }
}

// function addDebugLogs({
//   githubApiBaseUrlV4,
//   query,
//   variables,
//   githubResponse,
//   didThrow = false,
// }: {
//   githubApiBaseUrlV4: string;
//   query: DocumentNode;
//   variables?: Variables;
//   githubResponse: AxiosResponse<unknown>;
//   didThrow?: boolean;
// }) {
//   const gqlQueryName = getQueryName(query);
//   logger.info(
//     `POST ${githubApiBaseUrlV4} (name:${gqlQueryName}, status: ${
//       githubResponse.status
//     }${didThrow ? ', EXCEPTION THROWN' : ''})`,
//   );

//   logger.verbose(`Query: ${print(query)}`);
//   logger.verbose('Variables:', variables);
//   logger.verbose('Response headers:', githubResponse.headers);
//   logger.verbose('Response data:', githubResponse.data);
// }
