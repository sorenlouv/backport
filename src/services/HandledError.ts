import { Commit } from './sourceCommit/parseSourceCommit';

type ErrorContext =
  | {
      code: 'merge-conflict-exception';
      conflictingFiles: string[];
      commitsWithoutBackports: {
        formatted: string;
        commit: Commit;
      }[];
    }
  | {
      code: 'custom-exception';
      message: string;
    }
  | {
      code: 'no-branches-exception' | 'abort-exception';
    };

function getMessage(errorContext: ErrorContext): string {
  switch (errorContext.code) {
    case 'merge-conflict-exception':
      return `Commit could not be cherrypicked due to conflicts in: ${errorContext.conflictingFiles.join(
        ','
      )}`;
    case 'no-branches-exception':
      return 'There are no branches to backport to. Aborting.';
    case 'abort-exception':
      return 'Aborted';
    case 'custom-exception':
      return errorContext.message;
  }
}

export class HandledError extends Error {
  errorContext: ErrorContext;
  constructor(errorContextOrString: ErrorContext | string) {
    const errorContext: ErrorContext =
      typeof errorContextOrString === 'string'
        ? { code: 'custom-exception', message: errorContextOrString }
        : errorContextOrString;
    const message = getMessage(errorContext);
    super(message);
    Error.captureStackTrace(this, HandledError);
    this.name = 'HandledError';
    this.message = message;
    this.errorContext = errorContext;
  }
}
