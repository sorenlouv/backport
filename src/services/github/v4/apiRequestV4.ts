import axios from 'axios';
import { logger, logLevel } from '../../logger';
import { AxiosError } from 'axios';
import { HandledError } from '../../HandledError';

export interface GithubV4Response<DataResponse> {
  data: DataResponse;
  errors?: {
    type: string;
    path: string[];
    locations: {
      line: number;
      column: number;
    }[];
    message: string;
  }[];
}

export async function apiRequestV4<DataResponse>({
  githubApiBaseUrlV4,
  accessToken,
  query,
  variables,
  handleError = true,
}: {
  githubApiBaseUrlV4: string;
  accessToken: string;
  query: string;
  variables?: {
    [key: string]: string | number | null;
  };
  handleError?: boolean;
}) {
  logger.debug('Query (Github v4):', query);
  logger.debug('Variables (Github v4):', variables);

  try {
    const response = await axios.post<GithubV4Response<DataResponse>>(
      githubApiBaseUrlV4,
      { query, variables },
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${accessToken}`,
        },
      }
    );

    if (response.data.errors) {
      const newError = new Error();
      //@ts-ignore
      newError.response = response;
      throw newError;
    }

    logger.debug('Response headers (Github v4):', response.headers);
    logger.debug('Response data (Github v4)', response.data);

    return response.data.data;
  } catch (e) {
    if (logLevel !== 'debug') {
      logger.info('Query (Github v4):', query);
      logger.info('Variables (Github v4):', variables);
    }

    logger.info('Response headers (Github v4):', e.response?.headers);
    logger.info('Response data (Github v4)', e.response?.data);

    if (handleError) {
      throw handleGithubV4Error(e);
    }
    throw e;
  }
}

export function handleGithubV4Error(e: AxiosError<GithubV4Response<null>>) {
  // not github api error
  if (!e.response?.data) {
    return e;
  }

  const errorMessages = e.response.data.errors?.map((error) => error.message);
  if (errorMessages) {
    return new HandledError(`${errorMessages.join(', ')} (Github v4)`);
  }

  return new HandledError(
    `Unexpected response from Github API (v4):\n${JSON.stringify(
      e.response.data,
      null,
      2
    )}`
  );
}
