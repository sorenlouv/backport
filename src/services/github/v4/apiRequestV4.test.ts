import axios from 'axios';
import { HandledError } from '../../HandledError';
import { apiRequestV4 } from './apiRequestV4';

describe('apiRequestV4', () => {
  describe('when request succeeds', () => {
    let spy: jest.SpyInstance;
    let res: unknown;
    beforeEach(async () => {
      spy = jest.spyOn(axios, 'post').mockResolvedValue({
        data: {
          data: 'some data',
        },
      } as any);

      res = await apiRequestV4({
        accessToken: 'myAccessToken',
        githubApiBaseUrlV4: 'https://my-custom-api.com/graphql',
        query: 'myQuery',
        variables: {
          foo: 'bar',
        },
      });
    });

    it('should return correct response', async () => {
      expect(res).toEqual('some data');
    });

    it('should call with correct args', async () => {
      expect(spy).toHaveBeenCalledWith(
        'https://my-custom-api.com/graphql',
        {
          query: 'myQuery',
          variables: { foo: 'bar' },
        },
        {
          headers: {
            Authorization: 'bearer myAccessToken',
            'Content-Type': 'application/json',
          },
        }
      );
    });
  });

  describe('when request fails with error messages', () => {
    beforeEach(() => {
      jest.spyOn(axios, 'post').mockRejectedValue({
        response: {
          data: {
            errors: [
              { message: 'some error' },
              { message: 'some other error' },
            ],
          },
        },
      } as any);
    });

    it('should return parsed github error', async () => {
      return expect(
        apiRequestV4({
          accessToken: 'myAccessToken',
          githubApiBaseUrlV4: 'myApiHostname',
          query: 'myQuery',
          variables: {
            foo: 'bar',
          },
        })
      ).rejects.toThrowError(
        new HandledError(`some error, some other error (Github v4)`)
      );
    });
  });

  describe('when request fails without error messages', () => {
    beforeEach(() => {
      jest.spyOn(axios, 'post').mockRejectedValue({
        response: {
          data: {
            foo: 'bar',
          },
        },
      } as any);
    });

    it('should return parsed github error', async () => {
      return expect(
        apiRequestV4({
          accessToken: 'myAccessToken',
          githubApiBaseUrlV4: 'myApiHostname',
          query: 'myQuery',
          variables: {
            foo: 'bar',
          },
        })
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
