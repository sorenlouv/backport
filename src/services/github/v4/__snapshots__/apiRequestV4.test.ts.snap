// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiRequestV4 when request fails without error messages should return parsed github error 1`] = `
"Unexpected response from Github API (v4):
{
  \\"foo\\": \\"bar\\"
}"
`;
